File: 0042-trapping-rain-water/0042-trapping-rain-water.kt
class Solution {
    fun trap(height: IntArray): Int {
        var left = 0
        var right = height.size -1
        
        var leftMax = height[left]
        var rightMax = height[right]

        var answer = 0
        while (left < right){
            when {
                leftMax < rightMax -> {
                    answer += leftMax - height[left]
                    left++
                    leftMax = max(leftMax, height[left])
                }
                else -> {
                    answer += rightMax - height[right]
                    right--
                    rightMax = max(rightMax, height[right])
                }
            }
        }
        return answer
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        answer = 0
        left, right = 0, len(height)-1
        left_max, right_max = height[left], height[right]


        while left < right:
            if left_max < right_max:
                answer += left_max - height[left]
                left += 1
                left_max = max(left_max, height[left])
            else:
                answer += right_max - height[right]
                right -= 1
                right_max = max(right_max, height[right])
        return answer